librz/util/axml.c

```diff
@@ -0,0 +1,445 @@
+// SPDX-FileCopyrightText: 2021 keegan
+// SPDX-License-Identifier: LGPL-3.0-only
+
+#include <string.h>
+#include <rz_types.h>
+#include <rz_core.h>
+
+#include "axml_resources.h"
+
+enum {
+	TYPE_STRING_POOL = 0x0001,
+	TYPE_XML = 0x0003,
+	TYPE_START_NAMESPACE = 0x100,
+	TYPE_END_NAMESPACE = 0x101,
+	TYPE_START_ELEMENT = 0x0102,
+	TYPE_END_ELEMENT = 0x103,
+	TYPE_RESOURCE_MAP = 0x180,
+};
+
+enum {
+	RESOURCE_NULL = 0x00,
+	RESOURCE_REFERENCE = 0x01,
+	RESOURCE_STRING = 0x03,
+	RESOURCE_FLOAT = 0x04,
+	RESOURCE_INT_DEC = 0x10,
+	RESOURCE_INT_HEX = 0x11,
+	RESOURCE_BOOL = 0x12,
+};
+
+enum {
+	FLAG_UTF8 = 1 << 8,
+};
+
+// Beginning of every header
+RZ_PACKED(
+	typedef struct {
+		ut16 type;
+		ut16 header_size;
+		ut32 size;
+	})
+chunk_header_t;
+
+// String pool referenced throughout the Binary XML, there must only be ONE
+RZ_PACKED(
+	typedef struct {
+		ut32 string_count;
+		ut32 style_count;
+		ut32 flags;
+		ut32 strings_offset;
+		ut32 styles_offset;
+		ut32 offsets[];
+	})
+string_pool_t;
+
+RZ_PACKED(
+	typedef struct {
+		ut16 size;
+		ut8 unused;
+		ut8 type;
+		union {
+			ut32 d;
+			float f;
+		} data;
+	})
+resource_value_t;
+
+RZ_PACKED(
+	typedef struct {
+		ut32 namespace;
+		ut32 name;
+		ut32 unused;
+		resource_value_t value;
+	})
+attribute_t;
+
+RZ_PACKED(
+	typedef struct {
+		ut32 line;
+		ut32 comment;
+		ut32 namespace;
+		ut32 name;
+		ut32 flags;
+		ut16 attribute_count;
+		ut16 unused0;
+		ut16 unused1;
+		ut16 unused2;
+		attribute_t attributes[];
+	})
+start_element_t;
+
+RZ_PACKED(
+	typedef struct {
+		ut32 line;
+		ut32 comment;
+		ut32 namespace;
+		ut32 name;
+	})
+end_element_t;
+
+RZ_PACKED(
+	typedef struct {
+		ut32 line;
+		ut32 comment;
+		ut32 prefix;
+		ut32 uri;
+	})
+namespace_t;
+
+static char *string_lookup(string_pool_t *pool, const ut8 *data, ut64 data_size, ut32 i, size_t *length) {
```

Comment:
I don't like we have here a duplicate of what's inside `librz/util/str_search.c`. But it's out of the scope and probably worth opening an issue to suggest refactoring this.